src/pykeen/utils.py

```diff
@@ -1095,3 +1100,118 @@ def load_configuration(path: Union[str, pathlib.Path, os.PathLike]) -> Mapping[s
     import doctest
 
     doctest.testmod()
+
+
+def product_normalise(input_tensor: torch.FloatTensor) -> torch.FloatTensor:
+    r"""Normalise the input tensor along its embedding dimension so that the geometric mean is 1.
+
+    :param input_tensor: An input tensor with final dimension $d$.
+
+    :returns: An output tensor whose last order is normalized to have a geometric mean of 1.
+    """
+    step1_tensor = torch.abs(input_tensor)  # Compute absolute value of all entries
+    step2_tensor = step1_tensor.clamp(min=SANITY_EPSILON)  # Prevent zero values by adding a sanity epsilon
+    log_norm_tensor = torch.log(step2_tensor)  # Compute the log prior to computing the geom. mean
+    step3_tensor = torch.mean(log_norm_tensor, dim=-1, keepdim=True)
+    norm_volume = torch.exp(step3_tensor)
+    pre_norm_out = input_tensor / norm_volume
+    return pre_norm_out
+
+
+def compute_box(
+    base: torch.FloatTensor, delta: torch.FloatTensor, size: torch.FloatTensor
+) -> Tuple[torch.FloatTensor, torch.FloatTensor]:
+    r"""Compute the lower and upper corners of a resulting box.
+
+    :param base: The base position (box center) of the input relation embeddings.
+    :param delta: The base shape of the input relation embeddings.
+    :param size: The size scalar vectors of the input relation embeddings.
+    :return: Lower and upper bounds of the box whose embeddings are provided as input.
+    """
+    size_pos = torch.nn.functional.elu(size) + 1  # Enforce that sizes are strictly positive by passing through ELU
+    delta_norm = product_normalise(delta)  # Shape vector is normalized using the above helper function
+    delta_final = size_pos * delta_norm  # Size is learned separately and applied to normalized shape
+    # Product normalize the delta
+    first_bound = base - 0.5 * delta_final  # Compute potential boundaries by applying the shape in substraction
+    second_bound = base + 0.5 * delta_final  # and in addition
+    box_low = torch.minimum(first_bound, second_bound)  # Compute box upper bounds using min and max respectively.
+    box_high = torch.maximum(first_bound, second_bound)
+    return box_low, box_high
+
+
+def point_to_box_distance(
+    points: torch.FloatTensor, box_lows: torch.FloatTensor, box_highs: torch.FloatTensor
```

Comment:
trailing comma

Reply:
it is the will of `black` to not include trailing comma here

Reply:
You can add it, and it will not remove it (but introduce the line breaks we see at many other places) :confused: